### ISSUE TITLE
Parallel deadlock associated with meshes

### ISSUE THREAD

**Issue by jrmaddison** [firedrakeproject/firedrake#1569]
The following leads to a parallel deadlock on my machine, on iteration 2 (for 4 processes)

```
from firedrake import UnitSquareMesh
import mpi4py.MPI as MPI

n = 0
while True:
    n += 1
    print(f"Iteration {n:d} ...", end="", flush=True)
    UnitSquareMesh(20, 20)
    if MPI.COMM_WORLD.rank == 0:
        # Force desynchronization of the garbage collector
        class Empty:
            pass
        [Empty() for i in range(100000)]
    print(" done", flush=True)
```

**Comment by wence-** [comment]
So I have no idea how you would even go about trying to fix this. The refcounted part of the garbage collector is deterministic, but here you're deliberately (to demonstrate the bug) using the generational one.

Are there things that we can recommend users don't do that would mean firedrake objects are always refcount-collected?

**Comment by jrmaddison** [comment]
I don't understand why the above should deadlock, as there are no cyclic references to `mesh`.

Manually deleting the DMPlex avoids the deadlock:

```
from firedrake import UnitSquareMesh
import mpi4py.MPI as MPI

n = 0
while True:
    n += 1
    print(f"Iteration {n:d} ...", end="", flush=True)
    mesh = UnitSquareMesh(20, 20)
    # del mesh._topology._plex  # Uncommenting fixes the deadlock
    del mesh
    if MPI.COMM_WORLD.rank == 0:
        # Force desynchronization of the garbage collector
        class Empty:
            pass
        [Empty() for i in range(100000)]
    print(" done", flush=True)
```

The deadlock itself happens in the call `PETSc.DMPlex().createFromCellList` call in `firedrake.mesh._from_cell_list`.

**Comment by jrmaddison** [comment]
I might have misunderstood the cause of the bug, as the Empty objects also do not have circulate references, and the following does not deadlock

```
from firedrake import UnitSquareMesh
import mpi4py.MPI as MPI

n = 0
while True:
    n += 1
    print(f"Iteration {n:d} ...", end="", flush=True)
    mesh = UnitSquareMesh(20, 20)
    if MPI.COMM_WORLD.rank == 0:
        class Empty:
            pass
        [Empty() for i in range(100000)]
    del mesh
    print(" done", flush=True)
```

**Comment by wence-** [comment]
That latter thing is fine because when you just make the mesh you increase its refcount, and then when you call `del` you decrease the refcount. At that point the destructors are called and in a collective manner.

The problem occurs when the collection of the dead reference happens due to the generational mark and sweep collector, rather than the refcounting one: now you can no longer guarantee that it happens collectively.

**Comment by wence-** [comment]
OK, turns out I completely don;t understand how the python gc collection works.

**Comment by jrmaddison** [comment]
I believe there is a cyclic reference to `mesh._topology._plex`, in the `mesh._topology._callback` closure, which references the `add_overlap` closure, which references `self`.

However for me

```
mesh._topology._callback.__closure__[0].cell_contents()
```

causes Python to crash.

**Comment by wence-** [comment]
OK, I have a bunch, but the major parts are in PyOP2 I think, and that needs more thought. I have also not checked that I have not broken anything else.

Your example now works for me on:

wence/fix/circular-references in both firedrake and PyOP2.

I have not squashed all the circular refs in PyOP2: they occur in the ObjectCaching that we do.

**Comment by wence-** [comment]
I had a further go at this. It's a complete rat's nest unfortunately. I think it needs some sitting down and thinking hard about how (and where) we do caching of things in a way that doesn't introduce cyclic references.

The high-level problem at the moment is that (for efficiency) `FunctionSpace` objects are throwaway things, but avoid recreating heavy data by maintaining a cache of the cell-node-maps (etc...) on the mesh they were built on.

This causes issues because a mesh has a coordinate field, which is a function, which has a functionspace, which references a mesh.

Functionspaces, I think, have to hold a strong reference to a mesh, because otherwise this doesn't work:
```python
def make_space(...):
   mesh = ...
   return FunctionSpace(mesh, ...)
```

The question is how to ensure that the shared data, which is saved on the mesh, does not hold a strong reference to the mesh. and so on.

**Comment by JDBetteridge** [comment]
This is now fixed by #2599
